### PR TITLE
cudaPackages: eliminate exceptions

### PR DESCRIPTION
**Pull request by SomeoneSerge**
as they break common nixpkgs tools like lib.meta,
cf. https://github.com/NixOS/nixpkgs/issues/276795

@NixOS/cuda-maintainers 

## Description of changes

Before:

```
Before:
❯ nix repl --show-trace
Welcome to Nix 2.19.2. Type :? for help.
nix-repl> :lf nixpkgs
nix-repl> with legacyPackages.aarch64-darwin; lib.meta.availableOn "aarch64-darwin" cudaPackages.cuda_nvcc
...
       … while calling 'getRedistArch'
...
       error: Unsupported Nix system: aarch64-darwin
```

After:

```
nix-repl> with legacyPackages.aarch64-darwin; lib.meta.availableOn "aarch64-darwin" cudaPackages.cuda_nvcc
false
nix-repl> with legacyPackages.x86_64-darwin; cudaPackages.cuda_nvcc.meta.unsupported  
true
nix-repl> with legacyPackages.x86_64-linux; cudaPackages.cuda_nvcc.meta.unsupported  
false
```

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/development/cuda-modules/flags.nix [hunk 1]
```diff
@@ -175,9 +175,10 @@ let
 
       # archNames :: List String
       # E.g. [ "Turing" "Ampere" ]
+      #
+      # Unknown architectures are rendered as sm_XX gencode flags.
       archNames = lists.unique (
-        lists.map (cap: cudaComputeCapabilityToName.${cap} or (throw "missing cuda compute capability"))
-          cudaCapabilities
+        lists.map (cap: cudaComputeCapabilityToName.${cap} or "sm_${dropDot cap}") cudaCapabilities
```
Comment (SomeoneSerge) [inline pkgs/development/cuda-modules/flags.nix:181]: We can make this `warn "Unknown architecture: ${cap}" "sm_${...}"` later. Right now I just want cleaner eval

**Review by aidalgol — APPROVED** [review]

**Comment by SomeoneSerge** [comment]
Some of the outPaths have changed because the "unsupported" packages now have more `outputs`. Stuff like opencv4 and torch hasn't been affected, only the unsupported stuff has.

The solutions here are highly sub-optimal. There are upcoming refactors to the evalModules bit (incl. in Connor's cross-compilation PR) which might address this properly. It took some effort to satisfy Ofborg here without more global changes. 

Ofborg happy, and thanks to @aidalgol for the review. Merging this so we can start using `meta.unsupported`

File: pkgs/development/python-modules/tensorflow/default.nix [hunk 2]
```diff
@@ -480,6 +480,7 @@ let
     };
 
     meta = with lib; {
+      badPlatforms = lib.optionals cudaSupport lib.platforms.darwin;
```
Comment (SomeoneSerge) [inline pkgs/development/python-modules/tensorflow/default.nix:483]: This was required because of the "missing attribute cudnn_8_6"

**Comment by infinisil** [comment]
This indirectly broke the channel build, fix in https://github.com/NixOS/nixpkgs/pull/279684

**Pull request merged**